server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java

```diff
@@ -325,7 +326,17 @@ public DiscoveryNode(StreamInput in) throws IOException {
             }
         }
         this.roles = Collections.unmodifiableSortedSet(roles);
-        versionInfo = new VersionInformation(Version.readVersion(in), IndexVersion.readVersion(in), IndexVersion.readVersion(in));
+        Version version = Version.readVersion(in);
+        IndexVersion minIndexVersion = IndexVersion.readVersion(in);
+        IndexVersion minReadOnlyIndexVersion;
+        if (in.getTransportVersion().onOrAfter(NODE_VERSION_INFORMATION_WITH_MIN_READ_ONLY_INDEX_VERSION)) {
+            minReadOnlyIndexVersion = IndexVersion.readVersion(in);
```

Comment:
It is good, but perhaps you can confirm my thinking here. I originally worried a little about it persisting due to residing in cluster state such that it would depend on the master's version.

It should work though since we require a full upgrade to 8.18 before upgrading to 9.0 (at least in a rolling fashion). Hence on 8.18, any master here would then see a new 9.0 node, hence it would have the N-2 min-read-only-version.

All good I think, just overcomplicating it (I think).

Reply:
I think it is OK too, yet it is not an area I'm super confortable with.

My understanding is that a cluster on 8.17 with indices created in 7.x should reject join requests from 9.0 nodes since the node will only communicate a min. index compatible version on 8.

Once this PR is merged in `main` and backported to 8.18, a master on 8.18 will receive join requests from 9.0 nodes with both the N-1 and N-2 information but should still reject the node if 7.x indices exist in the cluster since 9.0 is not able to write such indices (I expect to change this soon in #118785).